File: examples/runnable/13-choreography/go/choreography.go
package choreography

import (
	"fmt"
	"time"

	"go.temporal.io/sdk/workflow"
)

const (
	TaskQueue  = "choreography"
	WorkflowID = "order-choreo-1001"
)

type DomainEvent struct {
	OrderID string
	Type    string
	Payload string
}

type StatusQuery struct {
	OrderID string
	Status  string
	Events  int
}

func OrderProcessWorkflow(ctx workflow.Context, orderID string) (string, error) {
	ctx = workflow.WithActivityOptions(ctx, workflow.ActivityOptions{
		StartToCloseTimeout: 10 * time.Second,
	})

	events := []DomainEvent{}
	status := "WAITING_FOR_ORDER"
	failureReason := ""
	signalChan := workflow.GetSignalChannel(ctx, "onEvent")

	state := func() string {
		return fmt.Sprintf("%s %s events=%d", orderID, status, len(events))
	}

	if err := workflow.SetQueryHandler(ctx, "status", func() (string, error) {
		return state(), nil
	}); err != nil {
		return "", err
	}

	receiveEvent := func() {
		var event DomainEvent
		signalChan.Receive(ctx, &event)
		events = append(events, event)
		if event.Type == "PaymentFailed" {
			failureReason = event.Payload
		}
	}

	hasEvent := func(eventType string) bool {
		for _, event := range events {
			if event.Type == eventType {
				return true
			}
		}
		return false
	}

	for !hasEvent("OrderPlaced") {
		receiveEvent()
	}
	status = "ORDER_ACCEPTED"
	if err := workflow.ExecuteActivity(ctx, ReserveLocalInventory, orderID).Get(ctx, nil); err != nil {
		return "", err
	}

	for !hasEvent("PaymentCaptured") && !hasEvent("PaymentFailed") {
		receiveEvent()
	}
	if hasEvent("PaymentFailed") {
		status = "COMPENSATING_PAYMENT_FAILURE"
		if err := workflow.ExecuteActivity(ctx, ReleaseInventory, orderID, failureReason).Get(ctx, nil); err != nil {
			return "", err
		}
		status = "CANCELLED"
		return state(), nil
	}

	status = "PAID_WAITING_FOR_INVENTORY"
	for !hasEvent("InventoryReserved") {
		receiveEvent()
	}
	if err := workflow.ExecuteActivity(ctx, RequestShipment, orderID).Get(ctx, nil); err != nil {
		return "", err
	}

	status = "READY_TO_SHIP"
	return state(), nil
}

func ReserveLocalInventory(orderID string) error {
	fmt.Println("reserved local inventory for", orderID)
	return nil
}

func RequestShipment(orderID string) error {
	fmt.Println("requested shipment for", orderID)
	return nil
}

func ReleaseInventory(orderID string, reason string) error {
	fmt.Printf("released inventory for %s: %s\n", orderID, reason)
	return nil
}

File: examples/runnable/13-choreography/python/choreography.py
from dataclasses import dataclass
from datetime import timedelta
from typing import List

from temporalio import activity, workflow

TASK_QUEUE = "choreography"
WORKFLOW_ID = "order-choreo-1001"


@dataclass
class DomainEvent:
    order_id: str
    type: str
    payload: str


@activity.defn
async def reserve_local_inventory(order_id: str) -> None:
    print(f"reserved local inventory for {order_id}")


@activity.defn
async def request_shipment(order_id: str) -> None:
    print(f"requested shipment for {order_id}")


@activity.defn
async def release_inventory(order_id: str, reason: str) -> None:
    print(f"released inventory for {order_id}: {reason}")


@workflow.defn
class OrderProcessWorkflow:
    def __init__(self) -> None:
        self._order_id = ""
        self._status = "WAITING_FOR_ORDER"
        self._events: List[DomainEvent] = []
        self._failure_reason = ""

    @workflow.run
    async def run(self, order_id: str) -> str:
        self._order_id = order_id

        await workflow.wait_condition(lambda: self._has_event("OrderPlaced"))
        self._status = "ORDER_ACCEPTED"
        await workflow.execute_activity(
            reserve_local_inventory,
            order_id,
            start_to_close_timeout=timedelta(seconds=10),
        )

        await workflow.wait_condition(
            lambda: self._has_event("PaymentCaptured")
            or self._has_event("PaymentFailed")
        )
        if self._has_event("PaymentFailed"):
            self._status = "COMPENSATING_PAYMENT_FAILURE"
            await workflow.execute_activity(
                release_inventory,
                args=[order_id, self._failure_reason],
                start_to_close_timeout=timedelta(seconds=10),
            )
            self._status = "CANCELLED"
            return self.status()

        self._status = "PAID_WAITING_FOR_INVENTORY"
        await workflow.wait_condition(lambda: self._has_event("InventoryReserved"))
        await workflow.execute_activity(
            request_shipment,
            order_id,
            start_to_close_timeout=timedelta(seconds=10),
        )

        self._status = "READY_TO_SHIP"
        return self.status()

    @workflow.signal
    def on_event(self, event: DomainEvent) -> None:
        self._events.append(event)
        if event.type == "PaymentFailed":
            self._failure_reason = event.payload

    @workflow.query
    def status(self) -> str:
        return f"{self._order_id} {self._status} events={len(self._events)}"

    def _has_event(self, event_type: str) -> bool:
        return any(event.type == event_type for event in self._events)

